File: app/src/main/java/com/example/liaopenghui/rxjavaabc/MainActivity.java
package com.example.liaopenghui.rxjavaabc;

import android.database.Observable;
import android.graphics.Bitmap;
import android.graphics.drawable.Drawable;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.util.Log;
import android.widget.ImageView;
import android.widget.Toast;

import com.example.liaopenghui.rxjavaabc.bean.Course;
import com.example.liaopenghui.rxjavaabc.bean.Student;

import java.util.ArrayList;
import java.util.Observer;
import java.util.logging.Logger;

import rx.Subscriber;
import rx.android.schedulers.AndroidSchedulers;
import rx.functions.Action0;
import rx.functions.Action1;
import rx.functions.Func1;
import rx.schedulers.Schedulers;

public class MainActivity extends AppCompatActivity {

    private static final String TAG = "MainActivity";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        /*********************基本用法和流程*******************/
        //    howToUseABC();
        //getName();
        //setDrawable();
        /********************scheduler的用法******************/
        //  practiceforSchedular();
        // practiceforSchedular1();
        /***********************最牛逼的变换********************/
        //  practiceForMap();
        practiceForFlatMap();
    }

    /**
     * 应用实例:
     * 现在要答应Students对象里的Course对象,并且一个Student有很多Course
     * 相当于发送(Observable)的是Student对象,得到的是Course对象  那么需要一个中转(中转的作用是发送Course对象)
     */
    private void practiceForFlatMap() {
        ArrayList<Student> students = getStudents();
        rx.Observable.from(students)
                .flatMap(new Func1<Student, rx.Observable<Course>>() {

                    @Override
                    public rx.Observable<Course> call(Student student) {
                        return rx.Observable.from(student.courses);
                    }
                }).subscribe(new Action1<Course>() {
            @Override
            public void call(Course course) {
                Log.e(TAG,"得到的课程有:"+course.courseName);
            }
        });
    }

    /**
     * 对map的联系  map 中  new Func1的参数:
     * String 发送的数据类型   和just()中的参数一致
     * bitMap   转换的数据类型   用户直接写  new Action1 中的泛型和它保持一致
     *
     */
//    private void practiceForMap() {
//        rx.Observable.just("images/logo.png")  //输入类型:String
//                .map(new Func1<String, Bitmap>() {
//                    @Override
//                    public Bitmap call(String filePath) {
//
//                     return getBitmapFromPath(filePath); // 返回类型 Bitmap
//                    }
//                }).subscribe(new Action1<Bitmap>() {
//            @Override
//            public void call(Bitmap bitmap) {
//                //接收到的类型 bitmap
//            }
//        });
////        final int drawableRes = R.mipmap.ic_launcher;
////        final ImageView iv_src = (ImageView) findViewById(R.id.iv_src);
////        rx.Observable.just(drawableRes)
////                .map(new Func1<Integer, Drawable>() {
////                    @Override
////                    public Drawable call(Integer drawable) {
////
////                        return getResources().getDrawable(drawable);
////                    }
////                }).subscribe(new Action1<Drawable>() {
////            @Override
////            public void call(Drawable bitmap) {
////                iv_src.setImageDrawable(bitmap);
////            }
////        });
//    }


    /**
     * 显示图片的就适合用scheduler进行  大图片的加载放在后台   线上放在mainThread上面
     */
    private void practiceforSchedular1() {
        final int drawableRes = R.mipmap.ic_launcher;
        final ImageView iv_src = (ImageView) findViewById(R.id.iv_src);
        rx.Observable.create(new rx.Observable.OnSubscribe<Drawable>() {
            @Override
            public void call(Subscriber<? super Drawable> subscriber) {
                Drawable drawable = getResources().getDrawable(drawableRes);
                subscriber.onNext(drawable);
                subscriber.onCompleted();
            }
        }).subscribeOn(Schedulers.io())   //io线程中加载图片
                .observeOn(AndroidSchedulers.mainThread())  //主线程中显示图片
                .subscribe(new Action1<Drawable>() {
                    @Override
                    public void call(Drawable drawable) {
                        iv_src.setImageDrawable(drawable);
                    }
                });
    }

    /**
     * 用来练习schedular
     */
    private void practiceforSchedular() {
        rx.Observable.just(1, 2, 3, 4)
                .subscribeOn(Schedulers.io())   //指定subscribe()发送在IO线程
                .observeOn(AndroidSchedulers.mainThread())//指定subscriber发送在主线程
                .subscribe(new Action1<Integer>() {
                    @Override
                    public void call(Integer integer) {
                        Log.e(TAG, "得到的number:" + integer);
                        Toast.makeText(MainActivity.this, "number" + integer, Toast.LENGTH_SHORT).show();
                    }
                });
    }

    private void setDrawable() {
        /******************************************************************实战演练:显示图片******************************************/
        final int drawableRes = R.mipmap.ic_launcher;
        final ImageView iv_src = (ImageView) findViewById(R.id.iv_src);
        rx.Observable.create(new rx.Observable.OnSubscribe<Drawable>() {
            @Override
            public void call(Subscriber<? super Drawable> subscriber) {
                Drawable drawable = getResources().getDrawable(drawableRes * 2);
                subscriber.onNext(drawable);
            }
        }).subscribe(new rx.Observer<Drawable>() {
            @Override
            public void onCompleted() {

            }

            @Override
            public void onError(Throwable e) {
                Log.e(TAG, "出错了");
                e.printStackTrace();
            }

            @Override
            public void onNext(Drawable drawable) {
                iv_src.setImageDrawable(drawable);
            }
        });
    }

    private void getName() {
        /******************************************************************实战演练:打印字符串数组******************************************/
        String[] names = {"zyy", "小公主", "廖鹏辉", "jay"};
        rx.Observable.from(names).subscribe(new Action1<String>() {
            @Override
            public void call(String s) {
                Log.e(TAG, "得到的名字为:" + s);
            }
        });
    }


    private void howToUseABC() {
        /******************************************************************第一步:接受数据的 ******************************************/
        rx.Observer<String> observer = new rx.Observer<String>() {

            @Override
            public void onCompleted() {
                Log.e(TAG, "Completed!observer");
            }

            @Override
            public void onError(Throwable e) {
                Log.e(TAG, "error!observer");
            }

            @Override
            public void onNext(String s) {
                Log.e(TAG, "ITEMobserver:" + s);
            }
        };

        //除了Observer接口之外,RxJava还内置了一个实现Observer的类:Subscriber
        Subscriber<String> subscriber = new Subscriber<String>() {

            @Override
            public void onCompleted() {
                Log.e(TAG, "Completed!subscriber");
            }

            @Override
            public void onError(Throwable e) {
                Log.e(TAG, "error!subscriber");
            }

            @Override
            public void onNext(String s) {
                Log.e(TAG, "ITEM:" + s);
            }
        };
        /******************************************************************第二步:发送数据的******************************************/
        rx.Observable<String> observable = rx.Observable.create(new rx.Observable.OnSubscribe<String>() {
            @Override
            public void call(Subscriber<? super String> subscriber) {
                subscriber.onNext("hello");
                subscriber.onNext("hi");
                subscriber.onNext("aloha");
                subscriber.onCompleted();
            }
        });


        rx.Observable<String> observable1 = rx.Observable.just("hello1", "hi1", "aloha");
        String[] arrys = {"hello2", "hi2", "aloha"};
        rx.Observable<String> observable2 = rx.Observable.from(arrys);


        /******************************************************************第三步订阅******************************************/
        observable.subscribe(subscriber);
        observable.subscribe(observer);
        observable1.subscribe(subscriber);
        observable1.subscribe(observer);
        observable2.subscribe(subscriber);
        observable2.subscribe(observer);


        /******************************************************************算是结束了******************************************/


        //除了 subscribe(Observer) 和 subscribe(Subscriber) ，subscribe() 还支持不完整定义的回调，RxJava 会自动根据定义创建出 Subscriber 。形式如下：

        Action1<String> onNextAction = new Action1<String>() {
            @Override
            public void call(String s) {
                Log.e(TAG, s);
            }
        };

        Action1<Throwable> onErrorAction = new Action1<Throwable>() {
            @Override
            public void call(Throwable throwable) {
                //错误回调
            }
        };

        Action0 onCompletedAction = new Action0() {
            @Override
            public void call() {
                Log.e(TAG, "completed");
            }

        };


        observable.subscribe(onNextAction);
        observable.subscribe(onNextAction, onErrorAction);
        observable.subscribe(onNextAction, onErrorAction, onCompletedAction);
    }

    public ArrayList<Student> getStudents() {
        ArrayList<Student> students = new ArrayList<>();
        for (int i = 0; i < 5; i++){
    Student student = new Student();
            ArrayList<Course> courseArrayList = new ArrayList<>();
    for(int p = 0;p<6;p++){
        Course course = new Course();
        course.courseName = "课程"+i+p;
        courseArrayList.add(course);
    }
            student.courses =courseArrayList;
            students.add(student);
        }

    return students;
}

}

File: app/src/main/java/com/example/liaopenghui/rxjavaabc/MainActivity.kt
package com.example.liaopenghui.rxjavaabc

import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.util.Log
import io.reactivex.*
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.disposables.Disposable
import io.reactivex.functions.BiFunction
import io.reactivex.observers.DisposableObserver
import io.reactivex.schedulers.Schedulers
import org.jetbrains.anko.AnkoLogger
import org.jetbrains.anko.info
import org.jetbrains.anko.toast
import java.util.concurrent.TimeUnit

class MainActivity : AppCompatActivity(), AnkoLogger {
    //mCompositeDisposable作用：如果一个页面有多个observer需要观察  可以通用放在里面
    //onDestory的时候 执行clear(),统一清理。
    val TAG = "MainActivity"
    val mCompositeDisposable: CompositeDisposable by lazy {
        CompositeDisposable()
    }
    lateinit var disposable: Disposable
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
//        baseFunc()
//        baseFunc1()
//        threadChange()
//        mapOprationSymble()
//        concatOperationSymble()
//        flatMapOprationSymble()
//        zipOprationSymble()
//        intervalOprationSymble()
//        debounceSymble()
//        takeSymble()
//        skipSymble()
//        throttleWithTimeoutSymble()
//        bufferSymble()
//        distinctSymble()
//        lastSymble()
//        reduceSymble()
        scanSymble()
    }


    private fun baseFunc() {
        Observable.create(ObservableOnSubscribe<Int> { emitter ->
            //第一步创建被观察者 Observable
            //emitter 发射器
            emitter.onNext(1)
            emitter.onNext(3)
            emitter.onNext(1)
            emitter.onNext(4)
            emitter.onNext(5)
            emitter.onNext(2)
            emitter.onNext(1)
        }).subscribe(object : Observer<Int> {
            //第二步创建观察者 Observer
            //第三步 订阅 subscribe
            override fun onComplete() {
                info { "onComplete" }
            }

            override fun onSubscribe(d: Disposable) {
                //当开始订阅的时候  提供了Disposable对象
                info { "onSubscribe" }
                disposable = d
            }

            override fun onNext(t: Int) {
                info { "onNext:$t" }
                toast("onNext:$t")
                if (t == 4) {
                    disposable.dispose()
                }
            }

            override fun onError(e: Throwable) {
            }

        })
    }

    private fun baseFunc1() {
        val subscribe = Observable.create(ObservableOnSubscribe<Int> { emitter ->
            //第一步创建被观察者 Observable
            //emitter 发射器
            emitter.onNext(1)
            emitter.onNext(3)
            emitter.onNext(1)
            emitter.onNext(4)
            emitter.onNext(5)
            emitter.onNext(2)
            emitter.onNext(1)
        }).subscribe { t -> info { "onNext:$t" } }
        mCompositeDisposable.add(subscribe)
    }

    //线程变换
    fun threadChange() {
        val subscribe = Observable.create(ObservableOnSubscribe<String> { emitter ->
            emitter.onNext("lph~~~")
            emitter.onNext("jay~~~")
            emitter.onNext("mayday~~~")
        }).subscribeOn(Schedulers.newThread())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe { t -> toast(t.toString()) }
        mCompositeDisposable.add(subscribe)
    }

    //map操作符
    fun mapOprationSymble() {
        var total: Int = 0
        val subscribe = Observable.create(ObservableOnSubscribe<String> { emitter ->
            emitter.onNext("5")
            emitter.onNext("5")
        }).map { t -> t.toInt() }.subscribe { t ->
            total += t!!
            info { "total:$total" }
        }

        mCompositeDisposable.add(subscribe)
    }

    //concat操作符
    fun concatOperationSymble() {
//        Observable.concat(getObservableA(null), getObservableB(null), getObservableA(null), getObservableB(null))
//            .subscribe(object : Consumer<String> {
//                override fun accept(t: String?) {
//                    info{t}
//                }
//
//            })


        var disposableObserver = object : DisposableObserver<String>() {
            override fun onNext(t: String) {
                info { "disposableObserver:$t" }
            }

            override fun onError(e: Throwable) {
            }

            override fun onComplete() {
            }

        }
        val subscribeWith =
            Observable.concat(getObservableA(null), getObservableB(null), getObservableA(null), getObservableB(null))
                .subscribeWith(disposableObserver)
        mCompositeDisposable.add(subscribeWith)
    }

    //concat操作符
    fun getObservableA(o: Any?): Observable<String> {
        return Observable.fromCallable {
            Thread.sleep(1000)//假设为耗时操作
            "liaopenghui"
        }
    }

    //concat操作符
    fun getObservableB(o: Any?): Observable<String> {
        return Observable.fromCallable {
            Thread.sleep(1000)//假设为耗时操作
            "jay jay jay"
        }
    }

    //flatMap 拿着上个函数的结果 作为下一个函数的参数
    fun flatMapOprationSymble() {
        val flatMapOprationSymbleSubscribe =
            Observable.create(ObservableOnSubscribe<String> { emitter -> emitter.onNext("123456789111") })
                .flatMap { t ->
                    if (t == "123456789") {
                        ObservableSource<String> { observer -> observer.onNext("密码正确") }
                    } else {
                        ObservableSource<String> { observer -> observer.onNext("密码错误") }
                    }
                }.subscribeOn(AndroidSchedulers.mainThread()).subscribe { t -> toast(t.toString()) }
        mCompositeDisposable.add(flatMapOprationSymbleSubscribe)
    }

    fun zipOprationSymble() {
        val zipSubscribe = Observable.zip(getObservableA(null), getObservableB(null),
            BiFunction<String, String, String> { t1, t2 -> t1 + t2 }).subscribeOn(AndroidSchedulers.mainThread())
            .subscribe { t -> toast(t.toString()) }
        mCompositeDisposable.add(zipSubscribe)
    }


    fun intervalOprationSymble() {
        disposable = Flowable.interval(1, TimeUnit.SECONDS).doOnNext { t -> error("accept: doOnNext : $t") }
            .subscribe { t -> error("accept: accept : $t") }
    }

    override fun onDestroy() {
        // 如果退出程序，就清除后台任务
        mCompositeDisposable.clear()
        disposable.dispose()
        super.onDestroy()
    }


    //debounce延迟操作符
    fun debounceSymble() {
        var debounceSymbleSubscribe = Observable.create(ObservableOnSubscribe<Int> { emitter ->
            emitter.onNext(1); // skip
            Thread.sleep(400);
            emitter.onNext(2); // deliver
            Thread.sleep(505);
            emitter.onNext(3); // skip
            Thread.sleep(100);
            emitter.onNext(4); // deliver
            Thread.sleep(605);
            emitter.onNext(5); // deliver
            Thread.sleep(510);
            emitter.onComplete()
        }).debounce(500, TimeUnit.MILLISECONDS).subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe { t -> toast(t.toString()) }
        mCompositeDisposable.add(debounceSymbleSubscribe)
    }

    fun takeSymble() {
        var subscribe = Observable.create(ObservableOnSubscribe<String> { emitter ->
            emitter.onNext("lph~~~")
            emitter.onNext("jay~~~")
            emitter.onNext("mayday~~~")
            emitter.onNext("mayday~1~~")
            emitter.onNext("mayday2~~~")
            emitter.onNext("mayday3~~~")
            emitter.onNext("mayday~4~~")
            emitter.onComplete()
        }).take(3).subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe { t -> Log.e(TAG, "" + t) }
        mCompositeDisposable.add(subscribe)


    }


    private fun skipSymble() {
        val subscribe =
            Observable.just(1, 2, 3, 4, 5, 6, 7).subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread())
                .skip(3).subscribe {
                    Log.e(TAG, "" + it)
                }
        mCompositeDisposable.add(subscribe)
        //output   4  5  6  7
    }


    //throttleWithTimeout操作符
    fun throttleWithTimeoutSymble() {
        var subscribe = Observable.create(ObservableOnSubscribe<Int> { emitter ->
            emitter.onNext(1); // skip
            Thread.sleep(400);
            emitter.onNext(2); // deliver
            Thread.sleep(505);
            emitter.onNext(3); // skip
            Thread.sleep(100);
            emitter.onNext(4); // deliver
            Thread.sleep(605);
            emitter.onNext(5); // deliver
            Thread.sleep(510);
            emitter.onComplete()
        }).throttleWithTimeout(500, TimeUnit.MILLISECONDS).subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe { t -> Log.e(TAG, "" + t) }
        mCompositeDisposable.add(subscribe)
    }


    private fun bufferSymble() {
        val subscribe = Observable.just("one", "two", "three", "four", "five")
            .buffer(3, 1) //每三个一缓冲,每次步进1
            .subscribe {
                //返回了一个List
                Log.e(TAG, "list:$it")
                //output:
//                 list:[one, two, three]
//                 list:[three, four, five]
//                 list:[two, three, four]
//                 list:[four, five]
//                 list:[five]
            }
        mCompositeDisposable.add(subscribe)
    }


    private fun distinctSymble() { //主要用于去重
        val subscribe = Observable.just(1, 2, 1, 1, 2, 3, 4, 6, 2, 1)
            .subscribeOn(Schedulers.io())
            .distinct()
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe {
                Log.e(TAG, "item:$it")
            }

        mCompositeDisposable.add(subscribe)
    }

    private fun lastSymble() {
        val subscribe = Observable.just("liao", "peng", "hui")
            .last("没人发送，就发送我自己")
            .subscribeOn(Schedulers.io())
            // Be notified on the main thread
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe { t ->
                {
                    Log.e(TAG, "item:$t")
                }
            }
        mCompositeDisposable.add(subscribe)
    }


    private fun reduceSymble() { //字面意思是减少  其实可以理解为聚合  把多个事件压缩 最后发射压缩后的事件
        val subscribe = Observable.just(1, 2, 1, 1, 2, 3, 4, 6, 2, 1)
            .subscribeOn(Schedulers.io())
            .reduce { t1: Int, t2: Int -> t1 + t2 }
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe {
                Log.e(TAG, "item:$it")  //最后对所有的数求和
            }

        mCompositeDisposable.add(subscribe)
    }

    private fun scanSymble() { //和reduce的用法差不多  但是每次处理的结果 它都会发射
        val subscribe = Observable.just(1, 2, 1, 1, 2, 3, 4, 6, 2, 1)
            .subscribeOn(Schedulers.io())
            .scan { t1: Int, t2: Int -> t1 + t2 }
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe {
                Log.e(TAG, "item:$it")  //最后对所有的数求和
            }

        mCompositeDisposable.add(subscribe)
    }

}